src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -324,55 +326,151 @@ public ProgrammingExercise setupProgrammingExercise(ProgrammingExercise programm
 
-            fileTargets.add("${exerciseNamePomXml}");
-            fileReplacements.add(programmingExercise.getTitle().replaceAll(" ", "-")); // Used e.g. in artifactId
+            replacePlaceholders(programmingExercise, repository);
+            pushRepository(repository, templateName);
+        }
+        else {
+            // If there is no special test structure for a programming language, just copy all the test files.
+            setupTemplateAndPush(repository, resources, prefix, templateName, programmingExercise);
+        }
+    }
 
-            fileTargets.add("${exerciseName}");
-            fileReplacements.add(programmingExercise.getTitle());
+    /**
+     * Replace placeholders in repository files (e.g. ${placeholder}).
+     * 
+     * @param programmingExercise
+     * @param repository
+     * @throws IOException
+     */
+    public void replacePlaceholders(ProgrammingExercise programmingExercise, Repository repository) throws IOException {
+        if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA) {
+            fileService.replaceVariablesInDirectoryName(repository.getLocalPath().toAbsolutePath().toString(), "${packageNameFolder}", programmingExercise.getPackageFolderName());
+        }
 
-            fileService.replaceVariablesInFileRecursive(repository.getLocalPath().toAbsolutePath().toString(), fileTargets, fileReplacements);
+        List<String> fileTargets = new ArrayList<>();
+        List<String> fileReplacements = new ArrayList<>();
+        // This is based on the correct order and assumes that boths lists have the same
+        // length, it
+        // replaces fileTargets.get(i) with fileReplacements.get(i)
 
-            gitService.stageAllChanges(repository);
-            gitService.commitAndPush(repository, templateName + "-Template pushed by Artemis");
-            repository.setFiles(null); // Clear cache to avoid multiple commits when ArTEMiS server is not restarted between attempts
+        if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA) {
+            fileTargets.add("${packageName}");
+            fileReplacements.add(programmingExercise.getPackageName());
         }
+        // there is no need in python to replace package names
+
+        fileTargets.add("${exerciseNamePomXml}");
+        fileReplacements.add(programmingExercise.getTitle().replaceAll(" ", "-")); // Used e.g. in artifactId
+
+        fileTargets.add("${exerciseName}");
+        fileReplacements.add(programmingExercise.getTitle());
+
+        fileService.replaceVariablesInFileRecursive(repository.getLocalPath().toAbsolutePath().toString(), fileTargets, fileReplacements);
+    }
+
+    /**
+     * Stage, commit and push.
+     * 
+     * @param repository
+     * @param templateName
+     * @throws GitAPIException
+     */
+    public void pushRepository(Repository repository, String templateName) throws GitAPIException {
```

Comment:
the method does more, so we should probably rename it in commitAndPush

Reply:
You are right!